src/strategies.rs

```diff
@@ -0,0 +1,195 @@
+//! This module contains an implementation of the `Hades252`
+//! strategy algorithm specifically designed to work outside of
+//! Rank 1 Constraint Systems (R1CS).
+//!
+//! The inputs of the permutation function have to be explicitly
+//! over the Scalar Field of the curve25519 so working over
+//! `Fp = 2^252 + 27742317777372353535851937790883648493`.
+use crate::mds_matrix::MDS_MATRIX;
+use crate::round_constants::ROUND_CONSTANTS;
+use crate::{PARTIAL_ROUNDS, TOTAL_FULL_ROUNDS};
+use bulletproofs::r1cs::{ConstraintSystem, LinearCombination};
+use curve25519_dalek::scalar::Scalar;
+use std::iter::FromIterator;
+use std::ops::Add;
+
+pub trait StrategyInput = From<Scalar> + Clone + Add;
+
+/// Defines the Hades252 strategy algorithm.
+pub trait Strategy<T>
+where
+    T: StrategyInput,
+    Vec<T>: FromIterator<<T as Add>::Output>,
+{
+    /// Computes `input ^ 5 (mod Fp)`
+    ///
+    /// The modulo depends on the input you use. In our case
+    /// the modulo is done in respect of the `curve25519 scalar field`
+    ///  == `2^252 + 27742317777372353535851937790883648493`.
+    fn quintic_s_box(&mut self, value: &T) -> T;
```

Comment:
We can't enforce the reallocation of the value since the mutation is performed on the value itself.

A suggestion is:
https://github.com/dusk-network/Hades252/pull/36/commits/41c5b800d336526122770b843787cb85ab6ef437

Reply:
_(Just for clarification for future reader, as discussed "we can" but "we shouldn't" in this case)._

Agreed, especially for the `LinearCombination` since it is on the heap. I will push the changes.

Please, as mentioned, edit the comment adding the relevant code snippet if you want (however the comment is already self-explained so it's up to you).